### PR TITLE
ANDROID: Allow to start a game directly from the Android launcher and various updates

### PR DESCRIPTION
**Pull request by lephilousophe**
Here is some big update.

I installed Android Studio to add the feature allowing to create shortcuts on the Android launcher and that dragged me in a series of changes before starting to work.
It contains :
- various small cleanups and style fixing (thanks to Android Studio),
- an update of translations with a fix on the translations helper script,
- a new Gradle properties file for the srcdir: this allows to have some static file copied in the build tree and a generated one for the moving part,
- I disabled Jetifier as we didn't need it (we use only androidx as a dependency and none other),
- various fixes in the dependency handling of the Makefile,
- a new INI file parser instead of hacking around the Java properties parser: this let us be sure we have the same results in Java and C++,
- add Path encoding parser to the new INI parser,
- some simple new features in SAFFSNode (to allow for use in Java side),
- allow to specify the game target inside the Intent (and handling of Activity restart).

Finally, I add the big feature.
The user can now add a shortcut from the Android launcher (long press, widgets button, scroll to ScummVM, select the Run a game widget).
The user gets a list of all games present in the configuration file (with a search field), each game has its icon if the user has downloaded the icon packs (like in C++, this works even if the icons are stored in SAF folders).

There is a catch: the official Java ZipFile implementation can only work on `File` objects (as it reopens the files when reading the compressed data). There is a `ZipInputStream` class which supports basic `InputStream` and which works well.
The problem is that this approach is really slow because each Zip file is fully read AND uncompressed (I think this could be optimized), so we have to read 300MB of data (icon packs) to display the icons of 1 or 2 games.

I solved this problem by taking the `ZipFile` class from Android and I patched it to accept `FileInputStream` objects. It's blazing fast but there is way more code. This approach is in the last two commits so this can be undoed if we decide to.
The changes I brought to the files imported from Android are prepended by a `ScummVM-changed` marker.
For this approach, the files are loaded when the game entry is displayed on the list and there is a cache (128 entries) if the user goes back in the list.

I think I covered everything I did here. :smile: 

The changes have been tested on Android 14, 11 (emulated) and 4.4 (KitKat).

### CODE REVIEW

File: backends/platform/android/org/scummvm/scummvm/ScummVMActivity.java [hunk 1]
```diff
@@ -839,6 +838,18 @@ protected String getScummVMLogPath() {
...
 
+		@Override
+		protected void setCurrentGame(String target) {
+			Uri data = null;
+			if (target != null) {
+				data = Uri.fromParts("scummvm", target, null);
+			}
+			Intent intent = new Intent(Intent.ACTION_MAIN, data,
+				ScummVMActivity.this, ScummVMActivity.class);
+//			Log.d(ScummVM.LOG_TAG, "Setting new Intent: " + data);
```
Comment (larsamannen) [inline backends/platform/android/org/scummvm/scummvm/ScummVMActivity.java:849]: I would keep this Log output

Comment (larsamannen) [reply backends/platform/android/org/scummvm/scummvm/ScummVMActivity.java:849]: Micro comment. Else LGTM

File: backends/platform/android/org/scummvm/scummvm/ScummVMActivity.java [hunk 2]
```diff
@@ -1063,6 +1075,64 @@ public void onStart() {
...
+		_finishing = true;
+
+		_events.clearEventHandler();
+		_events.sendQuitEvent();
+
+		// Make sure the thread is actively polling for events
+		_scummvm.setPause(false);
+		try {
+			// 1s timeout
+			_scummvm_thread.join(2000);
```
Comment (antoniou79) [inline backends/platform/android/org/scummvm/scummvm/ScummVMActivity.java:1114]: A small issue, but there is discrepancy between the comment (1s timeout) and the actual timeout in join (2000ms, ie. 2 seconcds)

**Comment by antoniou79** [comment]
I did a read through of the changes. In general, I think this is in good condition for merging and a really great effort (thank you!). 

The android cleanup stuff and the proper ini parser are easier for me to make sense of. 

The game shortcut code makes sense from a general look over, and I'd like it to be merged and tested by people who would use this in practice (I think I'm less likely to, but who knows, maybe I'll get hooked!)..

The zipUtils changes, I trust that they were made for efficiency and compatibility purposes as stated in the OP, and thus I can only offer few comments (mainly the one that stood out to me, which was the use of the legacy time libraries instead of the newer ones). Is this because the newer ones wouldn't work on old devices?

File: backends/platform/android/org/scummvm/scummvm/zip/ZipUtils.java [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+            /*
+            try {
+                LocalDateTime ldt = LocalDateTime.of(year, month, day, hour, minute, second);
+                return TimeUnit.MILLISECONDS.convert(ldt.toEpochSecond(
+                        ZoneId.systemDefault().getRules().getOffset(ldt)), TimeUnit.SECONDS);
+            } catch (DateTimeException dte) {
+                // ignore
+            }
+            */
+            return (new GregorianCalendar(year, month, day, hour, minute, second)).getTimeInMillis();
```
[the remark this answers is not in the document]
Comment (lephilousophe) [reply backends/platform/android/org/scummvm/scummvm/zip/ZipUtils.java:134]: As per [Android docs](https://developer.android.com/reference/java/time/LocalDateTime), LocalDateTime got added in Android 8 (API level 26).
Using the GregorianCalendar API is what was done in older versions of ZipUtils and works on older Android (nothing original by me).

File: backends/platform/android/org/scummvm/scummvm/ScummVMActivity.java (hunk 2 above)
Comment (lephilousophe) [reply backends/platform/android/org/scummvm/scummvm/ScummVMActivity.java:1114]: Thanks! Fixed in both places (as this piece of code was a copy-paste)

File: backends/platform/android/org/scummvm/scummvm/ScummVMActivity.java (hunk 1 above)
Comment (lephilousophe) [reply backends/platform/android/org/scummvm/scummvm/ScummVMActivity.java]: I changed the log message and commented it out although I am not really sure it's useful.

**Comment by lephilousophe** [comment]
> The zipUtils changes, I trust that they were made for efficiency and compatibility purposes as stated in the OP, and thus I can only offer few comments (mainly the one that stood out to me, which was the use of the legacy time libraries instead of the newer ones). Is this because the newer ones wouldn't work on old devices?

Indeed, it's to increase compatibility with old devices.
As we don't even use the time when reading our zip files, this piece of code is even not relevant but I tried to minimize the changes from the original.

Thank you for your review. As I rebased and the code still builds, I will merge it now.

**Pull request merged**